tls-core/src/prf/master.rs

```diff
@@ -1,225 +1,179 @@
+use super::errors::*;
 use super::sha::finalize_sha256_digest;
-use super::slave::{SlaveKe1, SlaveKe2, SlaveMs1, SlaveMs2};
 use super::utils::{seed_ke, seed_ms};
+use super::PRFMessage;
 
-pub struct Initialized;
-pub struct Ms1 {
-    /// master secret || client_random || server_random
-    seed: [u8; 77],
-    /// H(pms xor ipad)
-    inner_hash_state: [u32; 8],
-}
-
-pub struct Ms2 {
-    /// master secret || client_random || server_random
-    seed: [u8; 77],
-    /// H(pms xor ipad)
-    inner_hash_state: [u32; 8],
-}
-pub struct Ke1;
-
-pub struct Ke2 {
-    /// key expansion || client_random || server_random
-    seed: [u8; 77],
-    /// H(ms xor ipad)
-    inner_hash_state: [u32; 8],
-}
-
-pub struct Ke3 {
-    /// key expansion || client_random || server_random
-    seed: [u8; 77],
-    /// H(ms xor ipad)
-    inner_hash_state: [u32; 8],
-    /// H((ms xor opad) || H((ms xor ipad) || seed))
-    a1: [u8; 32],
-}
-
-pub struct Cf;
-
-pub trait State {}
-impl State for Initialized {}
-impl State for Ms1 {}
-impl State for Ms2 {}
-impl State for Ke1 {}
-impl State for Ke2 {}
-impl State for Ke3 {}
-impl State for Cf {}
-
-pub struct PrfMaster<S>
-where
-    S: State,
-{
-    /// State of 2PC PRF Protocol
-    state: S,
-    client_random: [u8; 32],
-    server_random: [u8; 32],
+#[derive(Copy, Clone)]
+pub struct MasterSetup {
+    pub inner_hash_state: [u32; 8],
 }
 
+#[derive(Copy, Clone)]
 pub struct MasterMs1 {
     /// H((pms xor ipad) || seed)
     pub inner_hash: [u8; 32],
 }
 
+#[derive(Copy, Clone)]
 pub struct MasterMs2 {
     /// H((pms xor ipad) || a1)
     pub inner_hash: [u8; 32],
 }
 
+#[derive(Copy, Clone)]
 pub struct MasterMs3 {
     /// H((pms xor ipad) || a2)
     pub inner_hash: [u8; 32],
 }
 
+#[derive(Copy, Clone)]
+pub struct KeInit {
+    pub inner_hash_state: [u32; 8],
+}
+
+#[derive(Copy, Clone)]
 pub struct MasterKe1 {
     /// H((ms xor ipad) || seed)
     pub inner_hash: [u8; 32],
 }
 
+#[derive(Copy, Clone)]
 pub struct MasterKe2 {
     /// H((ms xor ipad) || a1)
     pub inner_hash: [u8; 32],
 }
 
+#[derive(Copy, Clone)]
 pub struct MasterKe3 {
     /// H((ms xor ipad) || a1 || seed)
     pub inner_hash_p1: [u8; 32],
     /// H((ms xor ipad) || a2 || seed)
     pub inner_hash_p2: [u8; 32],
 }
 
-impl PrfMaster<Initialized> {
+#[derive(std::cmp::PartialEq, Copy, Clone)]
+pub enum PRFMasterState {
```

Comment:
This state will be mostly private (though we may want to expose it via a method), it's safe to be more succint. Eg. the enum can be named `State` and you don't need to prefix the variants. e.g
```
pub enum State {
  Initialized,
  Ms1,
  Ms2,
  Ms3
  ...
}
```